lib/src/rules/use_late.dart

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/features.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+
+import '../analyzer.dart';
+import '../util/dart_type_utilities.dart';
+
+const _desc = r'Use late for private members with non-nullable type.';
+
+const _details = r'''
+
+Use late for private members with non-nullable type to avoid null checks.
+
+**BAD:**
+```
+int? _i;
+m() {
+  _i!.abs();
+}
+```
+
+**GOOD:**
+```
+late int _i;
+m() {
+  _i.abs();
+}
+```
+
+''';
+
+class UseLate extends LintRule implements NodeLintRule {
+  UseLate()
+      : super(
+          name: 'use_late',
+          description: _desc,
+          details: _details,
+          maturity: Maturity.experimental,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends UnifyingAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  final lateables = <VariableDeclaration>[];
+  final nullableAccess = <AstNode, Element>{};
```

Comment:
I saw this earlier, but forgot to comment on it.

We don't really need to keep track of the element accessed at a given node, just just need to know, for each element, whether we have found a nullable access. Consider using the following instead:
```
final nullableAccess = <Element>{};
```
and just adding elements for which a nullable access was found to the set. Checking whether an element is "lateable" is then just a question of checking that it isn't in the set. That should keep this collection much smaller which will also improve performance.

Reply:
I keep the node to check the offset and evict every node at the same offset as the variable.

Reply:
The other (and in my opinion better) option would be to not add the elements associated with simple identifiers that are `inDeclarationContext()`.

Reply:
Done